### PR TITLE
Add actions to user context menu

### LINKED ISSUE #1674: Clicking on a username in a query window outputs userinfo again as new messages.
## Steps 
1. Click on a user in a regular window for which no query window is open.
2. After the query window opens click on the username. 

## Expected 
Nothing happens as we already have a query window open so the conversation is already ongoing. 

## Actual 
The user information is shown again. 

image

## Impact 

I'd say the impact is rather low, it mostly confused me on mobile as it is easier to accidentally click on a user in chat there than on desktop. 
When this happens in a channel and it opens up a window with the user it is clear what happend. 
In this case when that window is already open it just is confusing, I personally thought at first that I found a bug with how the backscroll is loaded.

### PR DESCRIPTION
**Pull request by creesch**
Resolves #1674.

This adds: 

- Whois
- Conversation 

And if the lounge user is also Op in the channel it adds: 

- Kick 
- Ban 

image

### CODE REVIEW

**Comment by creesch** [comment]
Added ab80d640fc988a26d9f1f20cf4cc2309bb5251a3 this fixes everything in #1674 as it will now show the context menu when you click a user instead of doing a `whois`.  As far as UX goes this is more in line of other IRC clients and should make The Lounge somewhat easier to use. 

On #thelounge there was some discussion about the ban `action` since it now, lacking better information,  defaults to banning based on nickname. I'll remove that in a future commit.

File: client/js/lounge.js [hunk 1]
```diff
@@ -72,6 +72,15 @@ $(function() {
 		return offset;
 	}
 
+	function isOpInChannel(channel) {
```
Comment (MaxLeiter) [inline client/js/lounge.js]: Could we move this to a new file? Doesn’t seen to fit into lounge.js

Comment (creesch) [reply client/js/lounge.js]: Move to utils.js maybe?

File: client/views/chan.tpl [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {{#each channels}}
-<div data-id="{{id}}" data-target="#chan-{{id}}" data-title="{{name}}" class="chan {{type}} chan-{{slugify name}}">
+<div data-id="{{id}}" data-target="#chan-{{id}}" data-title="{{name}}" class="chan {{type}} chan-{{slugify name}} chanid-{{id}}">
```
Comment (xPaw) [inline client/views/chan.tpl]: Why did you add this?

File: client/views/user_name.tpl [hunk 3]
```diff
@@ -1 +1 @@
-<span role="button" class="user {{colorClass nick}}" data-name="{{nick}}">{{mode}}{{nick}}</span>
+<span role="button" class="user {{colorClass nick}} nick-{{nick}}" data-name="{{nick}}">{{mode}}{{nick}}</span>
```
Comment (xPaw) [inline client/views/user_name.tpl]: And this?

File: client/views/chan.tpl (hunk 2 above)
Comment (creesch) [reply client/views/chan.tpl]: To keep [this function]( https://github.com/creesch/lounge/blob/ab80d640fc988a26d9f1f20cf4cc2309bb5251a3/client/js/lounge.js#L75-L82) from turning into a convoluted spaghetti mess. 
All the information I needed was already available but I needed a reliable way to cross reference it.

File: client/views/user_name.tpl (hunk 3 above)
Comment (creesch) [reply client/views/user_name.tpl]: Same as the other one.

File: client/views/chan.tpl (hunk 2 above)
Comment (xPaw) [reply client/views/chan.tpl]: You can already find it by `data-id`.

File: client/views/user_name.tpl (hunk 3 above)
Comment (xPaw) [reply client/views/user_name.tpl]: Same here, you can find it by `data-name`.

File: client/views/chan.tpl (hunk 2 above)
Comment (creesch) [reply client/views/chan.tpl]: That doesn't allow me to find it in all instances as jquery `data()` is used after the client is initialized. So when a nickname is changed it might not match the attribute anymore. 

For channels it appears to be relatively safe to use an attribute selector but I could not be certain so opted for including a more reliable selector in both instances.

**Comment by creesch** [comment]
@MaxLeiter moved the function to utils.js 
@xPaw changed the selectors to data-attributes as discussed on irc.

File: client/js/lounge.js [hunk 4]
```diff
@@ -442,15 +453,49 @@ $(function() {
 	});
 
 	contextMenu.on("click", ".context-menu-item", function() {
-		switch ($(this).data("action")) {
+		const $this = $(this);
+		const thisData = $this.data("data");
+		let chan;
+		switch ($this.data("action")) {
 		case "close":
```
Comment (YaManicKill) [inline client/js/lounge.js]: I feel like we should be starting to use the command pattern with these actions now rather than just increasing the case statement over time.

File: client/js/utils.js [hunk 5]
```diff
@@ -17,12 +17,22 @@ module.exports = {
...
 function resetHeight(element) {
 	element.style.height = element.style.minHeight;
 }
 
+// Given a channel element will determine if the lounge user is Op in that channel
+function isOpInChannel(channel) {
+	const channelID = channel.data("id");
+	const network = $("#sidebar .network").has(".chan[data-id=\"" + channelID + "\"]");
+	const ownNick = network.data("nick");
+	const isOP = channel.find(".users .user-mode.op .user[data-name=\"" + ownNick + "\"]").length;
```
Comment (xPaw) [inline client/js/utils.js]: Use string interpolation and `CSS.escape` (`const escape = require("css.escape");` for `ownNick`.

File: client/js/lounge.js [hunk 6]
```diff
@@ -82,6 +82,27 @@ $(function() {
...
 			});
+			output += templates.contextmenu_divider();
+			output += templates.contextmenu_item({
+				class: "action-whois",
+				text: "whois",
+				data: target.data("name"),
+			});
+			output += templates.contextmenu_item({
+				class: "action-query",
+				text: "conversation",
```
Comment (xPaw) [inline client/js/lounge.js]: Capitalise these strings.

File: client/js/lounge.js (hunk 4 above)
Comment (creesch) [reply client/js/lounge.js]: Can you elaborate, not sure what you mean by this.

Comment (YaManicKill) [reply client/js/lounge.js]: I don't really know what to say to elaborate.

Instead of having a case statement going through all of these actions, which can get very messy and long, we should have a command for each of them that are registered with a click action. These can live in separate places. If you don't know what the command pattern is, go [read up on it](https://en.wikipedia.org/wiki/Command_pattern). It's a very useful design pattern to know.

Comment (creesch) [reply client/js/lounge.js]: Ah, gotcha. Was indeed not familiar with it.  So something like this then?

```Javascript
var contextActions = {
    close: function(thisData) {
        $(`.networks .chan[data-target="${thisData}"] .close`).click();
    },
    chan: function(thisData) {
        $(`.networks .chan[data-target="${thisData}"]`).click();
    },
    whois: function(thisData) {
        const chan = utils.findCurrentNetworkChan(thisData);

        if (chan.length) {
            chan.click();
        }

        socket.emit("input", {
            target: $("#chat").data("id"),
            text: "/whois " + thisData,
        });

        $(`.channel.active .users .user[data-name="${thisData}"]`).click();
    },
    query: function(thisData) {
        const chan = utils.findCurrentNetworkChan(thisData);

        if (chan.length) {
            chan.click();
        }

        socket.emit("input", {
            target: $("#chat").data("id"),
            text: "/query " + thisData,
        });
    },
    kick: function(thisData) {
        socket.emit("input", {
            target: $("#chat").data("id"),
            text: "/kick " + thisData,
        });
    }
};

contextActions.execute = function(name) {
    return contextActions[name] && contextActions[name].apply(contextActions, [].slice.call(arguments, 1));
};
```
Somewhere tucked away. And then instead of the switch statement do something like

```Javascript

contextMenu.on("click", ".context-menu-item", function() {
    const $this = $(this);
    const thisData = $this.data("data");
    const contextAction = $this.data("action");
    contextActions.execute(contextAction, thisData)
});
```

Comment (YaManicKill) [reply client/js/lounge.js]: Almost, yeah. That would be a good start (probably good enough  for this PR) but at some point each of the contextActions should be in a separate file (1 for kick, 1 for query, etc). But as there aren't a huge number of them, probably not necessary yet, but wouldb egood at some point.

**Comment by creesch** [comment]
Did some searching and if accepted this would also fix #533 and partially #347.

**Comment by MaxLeiter** [comment]
@creesch can you squash commits? 👏

File: client/js/utils.js [hunk 7]
```diff
@@ -37,6 +39,15 @@ function resetHeight(element) {
 	element.style.height = element.style.minHeight;
 }
 
+// Given a channel element will determine if the lounge user is Op in that channel
+function isOpInChannel(channel) {
```
Comment (MaxLeiter) [inline client/js/utils.js]: I wonder if this should be `function isOpInChannel(user, channel)`?

Comment (creesch) [reply client/js/utils.js]: Doesn't make sense in the context of this commit as I specifically moved all of that (including the nick detection) to that function and would mean that I would have to move the first three lines back to [here](https://github.com/creesch/lounge/blob/46dfad0a0045ceede19718cc4859575a9675b7e6/client/js/lounge.js#L101)  where I explicitly removed it as to keep that area clean and readable.

**Comment by creesch** [comment]
Sure can do. 

image

Done :)

File: client/css/style.css [hunk 8]
```diff
@@ -220,12 +220,21 @@ kbd {
 #chat button.menu::before { content: "\f142"; /* http://fontawesome.io/icon/ellipsis-v/ */ }
 
 .context-menu-user::before { content: "\f007"; /* http://fontawesome.io/icon/user/ */ }
+
```
Comment (astorije) [inline client/css/style.css]: Could you remove this extra line?

File: client/css/style.css [hunk 9]
```diff
@@ -220,12 +220,21 @@ kbd {
...
+.context-menu-action-query::before { content: "\f0e6"; /* http://fontawesome.io/icon/comments-o/ */ }
+
+.context-menu-action-kick::before { content: "\f05e"; /* http://fontawesome.io/icon/ban/ */ }
+
+
 #sidebar .chan.lobby::before,
 #chat .lobby .title::before { content: "\f0a0"; /* http://fontawesome.io/icon/hdd-o/ */ }
 
+
```
Comment (astorije) [inline client/css/style.css]: Extra line here too.

File: client/css/style.css [hunk 10]
```diff
@@ -220,12 +220,21 @@ kbd {
...
 .context-menu-chan::before { content: "\f0f6"; /* http://fontawesome.io/icon/file-text-o/ */ }
 .context-menu-close::before { content: "\f00d"; /* http://fontawesome.io/icon/times/ */ }
 
+.context-menu-action-whois::before { content: "\f29c"; /* http://fontawesome.io/icon/question-circle-o/ */ }
+
+.context-menu-action-query::before { content: "\f0e6"; /* http://fontawesome.io/icon/comments-o/ */ }
+
+.context-menu-action-kick::before { content: "\f05e"; /* http://fontawesome.io/icon/ban/ */ }
+
+
```
Comment (astorije) [inline client/css/style.css]: Remove new lines between all those, also remove the one at R226.

File: client/js/lounge.js [hunk 11]
```diff
@@ -79,18 +79,45 @@ $(function() {
...
+			output += templates.contextmenu_item({
+				class: "action-whois",
+				action: "whois",
+				text: "Whois",
+				data: target.data("name"),
+			});
+			output += templates.contextmenu_item({
+				class: "action-query",
+				action: "query",
+				text: "Conversation",
```
Comment (astorije) [inline client/js/lounge.js]: Let's do "Send a direct message".

File: client/js/utils.js [hunk 12]
```diff
@@ -37,6 +39,15 @@ function resetHeight(element) {
...
+	const channelID = channel.data("id");
+	const network = $("#sidebar .network").has(`.chan[data-id="${channelID}"]`);
+	const ownNick = network.data("nick");
+	const isOP = channel.find(`.users .user-mode.op .user[data-name="${escape(ownNick)}"]`).length;
```
Comment (astorije) [inline client/js/utils.js]: @xPaw, isn't there a better way to determine if the current user is an op? If I read this correctly, this reads the info from the HTML of the user list.

File: client/js/lounge.js [hunk 13]
```diff
@@ -79,18 +79,45 @@ $(function() {
 		if (target.hasClass("user")) {
 			output = templates.contextmenu_item({
 				class: "user",
+				action: "whois",
 				text: target.text(),
 				data: target.data("name"),
 			});
+			output += templates.contextmenu_divider();
+			output += templates.contextmenu_item({
+				class: "action-whois",
```
Comment (astorije) [inline client/js/lounge.js]: Why are we duplicating this action? If this is for clarity sake (which would be a valid argument I think), then really we should make the first entry non-clickable. Visually different line for the name, à la Slack, minus the avatar picture:

screen shot 2017-12-10 at 23 43 08

File: client/js/lounge.js [hunk 14]
```diff
@@ -79,18 +79,45 @@ $(function() {
...
+				action: "whois",
+				text: "Whois",
```
Comment (astorije) [inline client/js/lounge.js]: "Information about this user"

File: client/js/lounge.js [hunk 15]
```diff
@@ -104,7 +131,11 @@ $(function() {
 		return false;
 	}
 
-	viewport.on("contextmenu", ".user, .network .chan", function(e) {
+	viewport.on("contextmenu", ".network .chan", function(e) {
+		return showContextMenu(this, e);
+	});
+
+	viewport.on("click contextmenu", ".user", function(e) {
```
Comment (astorije) [inline client/js/lounge.js]: I was very worried about the dropdown menu on left click, but it actually works very nicely. Nice job!
Especially like the fact that it's super snappy. On Slack it's not only slightly delayed but there is also an animation. Frustrating experience.

File: client/css/style.css [hunk 16]
```diff
@@ -220,12 +220,21 @@ kbd {
 #chat button.menu::before { content: "\f142"; /* http://fontawesome.io/icon/ellipsis-v/ */ }
 
 .context-menu-user::before { content: "\f007"; /* http://fontawesome.io/icon/user/ */ }
+
 .context-menu-chan::before { content: "\f0f6"; /* http://fontawesome.io/icon/file-text-o/ */ }
 .context-menu-close::before { content: "\f00d"; /* http://fontawesome.io/icon/times/ */ }
 
+.context-menu-action-whois::before { content: "\f29c"; /* http://fontawesome.io/icon/question-circle-o/ */ }
```
Comment (astorije) [inline client/css/style.css]: Could you use http://fontawesome.io/icon/info-circle/ for this instead please? (`f05a`)

File: client/js/lounge.js [hunk 17]
```diff
@@ -427,18 +444,56 @@ $(function() {
...
+		kick: function(itemData) {
+			socket.emit("input", {
+				target: $("#chat").data("id"),
+				text: "/kick " + itemData,
+			});
+		},
+	};
+
+	contextMenuActions.execute = function(name) {
+		return contextMenuActions[name] && contextMenuActions[name].apply(contextMenuActions, [].slice.call(arguments, 1));
```
Comment (astorije) [inline client/js/lounge.js]: Why the `.apply` and `[].slice.call`?
It seems to me you could add an `itemData` argument to the `execute` function and pass it to `contextMenuActions[name](itemData)`, no?

File: client/js/utils.js (hunk 12 above)
Comment (MaxLeiter) [reply client/js/utils.js]: this was discussed on IRC, that's basically the only way at the moment.

Comment (astorije) [reply client/js/utils.js]: Gotcha. If there is a better way that could be possible when refactoring this or that, do we want to add a TODO note here?

File: client/js/lounge.js (hunk 14 above)
Comment (creesch) [reply client/js/lounge.js]: I rather keep whois here, as that is the commonly known irc terminology.

File: client/js/lounge.js (hunk 11 above)
Comment (creesch) [reply client/js/lounge.js]: That is misleading, you are effectively starting a conversation which is different from direct message functionality imho.

File: client/js/lounge.js (hunk 13 above)
Comment (creesch) [reply client/js/lounge.js]: Left the duplicate activity because long time lounge users might expect that action. Also it doesn't hurt.

File: client/js/lounge.js (hunk 11 above)
Comment (astorije) [reply client/js/lounge.js]: What about "Direct messages" then (like Slack)? "Conversation" kind of means everything and anything is my issue.

File: client/js/lounge.js (hunk 14 above)
Comment (astorije) [reply client/js/lounge.js]: And I'd... rather not 😅
New users would have no idea what this is, while experienced users would feel free to type `/whois` anyway.
You could also do "Information about this user (`whois`)" if you reaaally want to keep "whois" in there. I'd be okay with that.

File: client/js/lounge.js (hunk 13 above)
Comment (astorije) [reply client/js/lounge.js]: Well, it's going literally 5mm lower, so it's not too confusing either :)
I think we should explore better UI ideas for this, but it doesn't have to be done in your PR.

File: client/js/lounge.js (hunk 17 above)
Comment (creesch) [reply client/js/lounge.js]: Simply future prooving and OOP oriented. If the handling method would change it would require all objects directly accessing these methods  to also be modified. By abstracting it away further like this we accept any named methods that can be performed on the object, passing along any data that might be used.

File: client/js/lounge.js (hunk 11 above)
Comment (creesch) [reply client/js/lounge.js]: How about "private conversation"?

Comment (astorije) [reply client/js/lounge.js]: "Conversation" still carries this confusing meaning. I prefer "messages" over "conversation". "Private" or "Direct" work, though "private" kinda gives this vibe or privacy, i.e. OTR or something, which we don't do, so I prefer "Direct", but I'm fine with either.

File: client/js/lounge.js (hunk 17 above)
Comment (astorije) [reply client/js/lounge.js]: Could you try this? It might work and is more explicit/simpler (also FP > OOP 😁):

```js
contextMenuActions.execute = (name, ...args) => contextMenuActions[name] && contextMenuActions[name](args);
```

Comment (creesch) [reply client/js/lounge.js]: Hrm I guess that could work, didn't think of using a rest parameter.

Comment (astorije) [reply client/js/lounge.js]: Actually, that would be `contextMenuActions[name](...args)` too, rather.

File: client/js/lounge.js (hunk 14 above)
Comment (creesch) [reply client/js/lounge.js]: Would you be okay with just "User information"?  Fits much better with the width of the menu.

Comment (YaManicKill) [reply client/js/lounge.js]: How is "whois" confusing? It's literally saying it in the name what it is.

I'm really not a fan of these long sentences you want in the context menu.

How about "Who is this user?" or something? Then you get the whois in, but it's a bit more of a sentence for new users.

File: client/js/lounge.js (hunk 11 above)
Comment (YaManicKill) [reply client/js/lounge.js]: What's the problem with Conversation?

File: client/js/lounge.js (hunk 14 above)
Comment (YaManicKill) [reply client/js/lounge.js]: (I'd definitely be fine with "User Information", that works fine for me)

File: client/js/lounge.js (hunk 11 above)
Comment (creesch) [reply client/js/lounge.js]: Still curious about this as well, for all intends and purposes you are starting a conversation as it does open the query window. Direct message more or less implies a fire and forget sort of thing more like mail or private messaging on forums. 

I don't care too strongly either way, matter of fact already changed it but am still curious.

**Comment by creesch** [comment]
@astorije should be all of it. Just need to squash stuff....

File: client/css/style.css [hunk 18]
```diff
@@ -224,6 +224,9 @@ kbd {
 .context-menu-chan::before { content: "\f292"; /* http://fontawesome.io/icon/hashtag/ */ }
 .context-menu-close::before { content: "\f00d"; /* http://fontawesome.io/icon/times/ */ }
 .context-menu-list::before { content: "\f03a"; /* http://fontawesome.io/icon/list/ */ }
+.context-menu-action-whois::before { content: "\f05a"; /* http://fontawesome.io/icon/info-circle/  */ }
+.context-menu-action-query::before { content: "\f0e6"; /* http://fontawesome.io/icon/comments-o/ */ }
+.context-menu-action-kick::before { content: "\f05e"; /* http://fontawesome.io/icon/ban/ */ }
```
Comment (astorije) [inline client/css/style.css]: FYI, I just removed an extra line L237. No big deal, I just didn't want to bother you again with something minor.

File: client/js/lounge.js (hunk 11 above)
Comment (astorije) [reply client/js/lounge.js]: It was 2am so maybe I wasn't clear, sorry :)
Issue with this is that "Conversation" is too broad. Actually, let me give you what I expected: when I first ran your PR, I had not looked at the code yet, I wanted to see the look-and-feel without a bias first. When I saw "Conversation", seeing that there was also "Whois" to open a window (with whois information, granted), I was totally expecting that it would bring the nick to the current conversation by auto-completing its name (it's not such a far-fetched idea, it had been brought up by a few people in the past).
I was miles away expecting it would open a new window without the whois :man_shrugging:

File: client/js/lounge.js (hunk 14 above)
Comment (astorije) [reply client/js/lounge.js]: User information is great, thanks! And yeah, "whois" doesn't say much unless you already know what a whois is in the context of IRC. "User information" (I was also fine with "User information (whois)") is way clearer :D

Re: sentences, I totally understand. It's difficult to give a proper explanation in 3 words max, but yeah my suggestion was too long. Thanks all!
